File: old-projects/fireworks-silverlight-port/original-csharp-source/BaseSprite.cs
using System.Windows.Media.Imaging;

namespace Fireworks_Silverlight
{
    public abstract class BaseSprite
    {
        protected double x, y, dx, dy, oldx, oldy;

        public abstract bool Animate(double height, double width, double multiplier);
        
        public abstract void Render(WriteableBitmap bmp, int h, int w, int[] pixels);

        public abstract void Render(WriteableBitmap bmp);

        protected void UpdatePosition(double multiplier)
        {
            var ndy = dy * multiplier;
            var ndx = dx * multiplier;
            y += ndy; // move
            x += ndx;
        }
    }
}

File: silverlight-ports/fireworks/src/BaseSprite.ts
export abstract class BaseSprite {
  protected x = 0;
  protected y = 0;
  protected dx = 0;
  protected dy = 0;
  protected oldx = 0;
  protected oldy = 0;

  public abstract Animate(height: number, width: number, multiplier: number): boolean;

  public abstract RenderShapes(context: CanvasRenderingContext2D): void;

  protected UpdatePosition(multiplier: number): void {
    const ndy = this.dy * multiplier;
    const ndx = this.dx * multiplier;
    this.y += ndy;
    this.x += ndx;
  }
}
